File: ios/RNMBTileSource.swift
//
//  RNMBTileSource.swift
//  ForestWatcher
//
//  Created by Ryan Bourne on 14/05/2020.
//  Copyright © 2020 3 SIDED CUBE. All rights reserved.
//

import Foundation
import SQLite
import UIKit

public struct RNMBTileMetadata {
  let minZoomLevel: Int
  let maxZoomLevel: Int
  let isVector: Bool
  let tms: Bool
  let tileSize: Int
  let attribution: String?
  let layersJson: String?
}

// MARK: MbtilesSource
enum MBTilesSourceError: Error {
  case CouldNotReadFileError
  case UnknownFormatError
  case UnsupportedFormatError
}

/// Defines a MBTileSource object, which is used by the tileserver to query for given tile indexes.
@objc public class RNMBTileSource: NSObject {
  
  // MBTiles spec, including details about vector tiles:
  // https://github.com/mapbox/mbtiles-spec/
  
  /// Defines the raster formats the app is supporting.
  static let validRasterFormats = ["jpg", "png"]
  
  /// Defines the vector formats the app is supporting.
  static let validVectorFormats = ["pbf", "mvt"]
  
  /// The basemap's unique identifier. When filtering requests, we use this to ensure the request is for this basemap.
  var id: String
  var filePath: String
  var metadata: RNMBTileMetadata?
  
  private var database: Connection?
  
  /// Initialises a source object with the given path and identifier.
  /// - Parameters:
  ///   - id: The basemap's unique identifier.
  ///   - filePath: The basemap's filepath on disk.
  /// - Throws: An error if the provided inputs do not result in a valid database connection, or if the provided basemap is invalid.
  public init(id: String, filePath: String) throws {
    self.id = id
    self.filePath = filePath
    
    super.init()
    
    database = try Connection(filePath, readonly: true)
    
    guard let anyTile = try database?.scalar("SELECT tile_data FROM tiles LIMIT 1") as? Blob else {
      throw MBTilesSourceError.CouldNotReadFileError
    }
    
    let tileData = Data(anyTile.bytes)
    
    let minZoom = Int(getMetadata(fieldName: "minzoom") ?? "0")
    let maxZoom = Int(getMetadata(fieldName: "maxzoom") ?? "0")
    
    // https://stackoverflow.com/a/42104538
    let headerData = [UInt8](tileData)[0]
    let format = getFormat(for: headerData)
    let isVector = try isVectorFormat(basedOn: format)
    
    let tileSize = getTileSize(basedOn: tileData)
    let attribution = getMetadata(fieldName: "attribution")
    let layersJson = getMetadata(fieldName: "json")
    
    metadata = RNMBTileMetadata(minZoomLevel: minZoom ?? 0, maxZoomLevel: maxZoom ?? 0, isVector: isVector, tms: true, tileSize: tileSize, attribution: attribution, layersJson: layersJson)
  }
  
  /// Determines the format of the basemap files.
  /// - Parameter headerData: The first data byte from a tile object.
  /// - Returns: The format of the basemap tile.
  private func getFormat(for headerData: UInt8) -> String? {
    if headerData == 0x89 {
      return "png"
    } else if headerData == 0xFF {
      return "jpg"
    } else {
      return getMetadata(fieldName: "format")
    }
  }
  
  /// Determines if the basemap uses vector or raster tiles. We use this in the basemap metadata, to tell the JS layer how to render the given tiles.
  /// - Parameter format: The pre-determined basemap format.
  /// - Throws: An error if the format is not provided, or if the format is not recognised.
  /// - Returns: If the basemap tiles are vector-based or rasterised.
  private func isVectorFormat(basedOn format: String?) throws -> Bool {
    guard let format = format else {
      throw MBTilesSourceError.UnknownFormatError
    }
    
    if RNMBTileSource.validRasterFormats.contains(format) {
      return false
    } else if RNMBTileSource.validVectorFormats.contains(format) {
      return true
    } else {
      throw MBTilesSourceError.UnsupportedFormatError
    }
  }
  
  /// Gets the basemap tile size.
  /// - Parameter tileData: A basemap tile.
  /// - Returns: The size of the tile (presuming both width & height to match).
  private func getTileSize(basedOn tileData: Data) -> Int {
    guard let tileImage = UIImage(data: tileData) else {
      return 0
    }
    
    let screenScale = UIScreen.main.scale > 1 ? 2 : 1
    return Int(tileImage.size.height / CGFloat(screenScale))
  }
  
  private func getMetadata(fieldName: String) -> String? {
    let query = "SELECT value FROM metadata WHERE name=\"\(fieldName)\""
    if let binding = try? database?.scalar(query) {
      return binding as? String
    }
    return nil
  }
  
  public func getTile(x: Int, y: Int, z: Int) -> Data? {
    let query = "SELECT tile_data FROM tiles WHERE tile_column=\(x) AND tile_row=\(y) AND zoom_level=\(z)"
    if let binding = try? database?.scalar(query),
      let blob = binding as? Blob {
      return Data(blob.bytes)
    }
    return nil
  }
}

File: react-native-mbtiles/android/src/main/java/com/reactlibrary/RNMBTileSource.kt
package com.forestwatcher

import android.database.sqlite.SQLiteDatabase
import com.facebook.react.bridge.WritableNativeMap
import org.jetbrains.anko.db.MapRowParser
import org.jetbrains.anko.db.select

// Defines metadata for this source object.
// TODO: Pass this up to the JS layer.
data class RNMBTileMetadata(
    var minZoomLevel: Int,
    var maxZoomLevel: Int,
    var isVector: Boolean,
    var tms: Boolean,
    var tileSize: Int,
    var attribution: String?,
    var layersJson: String?
)
{

    val mappedMetadata: WritableNativeMap
        get() {
            var nativeMap: WritableNativeMap = WritableNativeMap()
            nativeMap.putInt("minZoomLevel", minZoomLevel)
            nativeMap.putInt("maxZoomLevel", maxZoomLevel)
            nativeMap.putBoolean("isVector", isVector)
            nativeMap.putBoolean("tms", tms)
            nativeMap.putInt("tileSize", tileSize);
            nativeMap.putString("attribution", attribution)
            nativeMap.putString("layersJson", layersJson)

            return nativeMap
        }
}

// Defines various errors that may occur whilst working against this source.
sealed class RNMBTileSourceError : Error() {
    class CouldNotReadFileError : RNMBTileSourceError()
    class TileNotFoundError : RNMBTileSourceError()
    class UnsupportedFormatError : RNMBTileSourceError()
}

// Defines a MBTileSource object, which is used by the tileserver to query for given tile indexes.
class RNMBTileSource(var id: String, var filePath: String) {

    companion object {
        /// Defines the raster formats the app is supporting
        val VALID_RASTER_FORMATS = listOf("jpg", "png")

        /// Defines the vector formats the app is supporting
        val VALID_VECTOR_FORMATS = listOf("pbf", "mvt")
    }

    // Defines metadata for this source object.
    var metadata: RNMBTileMetadata? = null

    // These two properties are used in the server for constructing response payloads.
    var format: String? = null
    var isVector: Boolean = false

    // Defines a connection to the mbtiles database.
    // TODO: Refactor this to use a non-deprecated database dependency.
    private val database: SQLiteDatabase = try {
        SQLiteDatabase.openOrCreateDatabase(filePath, null)
    } catch (e: RuntimeException) {
        throw RNMBTileSourceError.CouldNotReadFileError()
    }

    // Initialises a source object with the given path and identifier.
    // We make various requests against the database, and pull out relevant metadata.
    // This metadata can then be used within the JS layer to correctly render tiles.
    init {
        try {
            val minZoomLevel = getMetadata("minzoom")?.toInt() ?: 0
            val maxZoomLevel = getMetadata("maxzoom")?.toInt() ?: 0
            format = getMetadata("format")
            val tms = true
            isVector = when (format) {
                in VALID_VECTOR_FORMATS -> true
                in VALID_RASTER_FORMATS -> false
                else -> throw RNMBTileSourceError.UnsupportedFormatError()
            }
            val tileSize = 256
            val attribution = getMetadata("attribution")
            val layersJson = null// getMetadata("json")

            metadata = RNMBTileMetadata(minZoomLevel = minZoomLevel, maxZoomLevel = maxZoomLevel, isVector = isVector, tms = tms, tileSize = tileSize, attribution = attribution, layersJson = layersJson)
        } catch (error: RNMBTileSourceError) {
            print(error.localizedMessage)
            throw error
        }
    }

    // Given coordinates, queries the database and returns a tile if one exists.
    fun getTile(z: Int, x: Int, y: Int): ByteArray {
        return database.select("tiles")
                .whereArgs("(zoom_level = {z}) and (tile_column = {x}) and (tile_row = {y})",
                        "z" to z, "x" to x, "y" to y)
                .parseList(TilesParser)
                .run { if (!isEmpty()) get(0) else null }
                ?: throw RNMBTileSourceError.TileNotFoundError()
    }

    // Given a metadata property, queries the database and returns it, if it exists.
    private fun getMetadata(property: String): String? {
        return database.select("metadata").whereSimple("name = ?", property).parseOpt(MetadataParser)?.second
    }
}

object MetadataParser : MapRowParser<Pair<String, String>> {
    override fun parseRow(columns: Map<String, Any?>): Pair<String, String> =
            columns["name"] as String to columns["value"] as String
}

object TilesParser : MapRowParser<ByteArray> {
    override fun parseRow(columns: Map<String, Any?>): ByteArray = columns["tile_data"] as ByteArray
}
